Customer return

aleph1ow opened the pull request:
Posts a return to Shipwire after saving a CustomerReturn.
Supports returning multiple line items with quantity.

DanielePalombo committed:
Rubocop fix

DanielePalombo committed:
Fix rspec test for solidus v1.0

DanielePalombo committed:
Remove DEPRECATION warning for rails 5.0

DanielePalombo committed:
Test more solidus branches

DanielePalombo committed:
Test more solidus branches

DanielePalombo committed:
Remove Gemfile.lock

DanielePalombo committed:
modify travis file email notification

DanielePalombo committed:
Add solidus versions < 1.4

DanielePalombo committed:
Fix specs for solidus version <= 1.1

DanielePalombo committed:
Merge pull request #6 from nebulab/solidusio-contrib

Prepare to solidusio-contrib

aleph1ow committed:
Add basic CustomerReturn decorator

aleph1ow committed:
Use to_shipwire instead of passing parameter

aleph1ow committed:
Add proper proxied customer return model

aleph1ow committed:
Display flash msg when adding customer return

aleph1ow committed:
Fix rescue indentation

aleph1ow committed:
Add tests for shipwire returns

DanielePalombo commented on app/models/customer_return_decorator.rb:
```diff
@@ -0,0 +1,83 @@
+module SolidusShipwire::CustomerReturn
```
Can you nest these modules?

DanielePalombo commented on app/models/customer_return_decorator.rb:
```diff
@@ -0,0 +1,83 @@
+    {
+      externalId: order.number,
+      originalOrder: {
+        id: order.shipwire_id.to_s
```
`to_s` is required here?

DanielePalombo commented on app/models/customer_return_decorator.rb:
```diff
@@ -0,0 +1,83 @@
+  end
+
+  def process_shipwire_return!
+  create_on_shipwire
```
nest this please

DanielePalombo commented on app/models/customer_return_decorator.rb:
```diff
@@ -0,0 +1,83 @@
+  def process_shipwire_return!
+  create_on_shipwire
+  rescue SolidusShipwire::ResponseException => e
+    if e.response.validation_errors
```
if this condition is false we don't do anything?

aleph1ow replied:
If the condition is false I do nothing. This is the case of a rare Timeout exception. All the other exceptions have validation errors and I forward them to the object. Timeout does not have any validation errors attached. How should I rescue from it?

DanielePalombo commented on app/models/customer_return_decorator.rb:
```diff
@@ -0,0 +1,83 @@
+    SolidusShipwire::ShipwireObjects::Return.new(hash['id'], self, hash)
+  end
+
+  def shipwire_line_items
```
This method `shipwire_line_items` is related to `return_items`, maybe can be better rename it in `shipwire_return_items`?

DanielePalombo commented on spec/models/spree/customer_return_spec.rb:
```diff
@@ -0,0 +1,128 @@
+describe Spree::CustomerReturn, type: :model do
+
+  # => Actually, just in case you want to regenerate vcr you must manually
```
This logic is very complex! Do you think can be created an helper to do these steps and avoid human errors?

aleph1ow replied:
There is a TODO annotation that describes how to generate orders in shipwire with particular state via shipwire API. Actually I don't know if this is possible but I've left the instructions for a future implementation. I could move these notes to a new issue. What do you think about it @DanielePalombo ?

DanielePalombo reviewed (commented):
great PR , I created only some note.

kennyadsl commented on spec/models/spree/customer_return_spec.rb:
```diff
@@ -0,0 +1,128 @@
+  # => 4 -  Create a complete order that contains 1 line item with variant with sku_1, quantity: 2,
+  # =>      1 line item with variant with sku_2, quantity: 2
+  # =>      and after save mark it as Shipped on shipwire (via api)
+  # => 3 -  Create a complete order that contains 1 line item with variant with sku_1
```
why 3 after 4 ?

kennyadsl commented on spec/models/spree/customer_return_spec.rb:
```diff
@@ -0,0 +1,128 @@
+          customer_return.order.update_attribute(:shipwire_id, shipwire_id)
+        end
+
+        it 'will post to shipwire' do
```
please use `posts` instead of `will post`

kennyadsl commented on spec/models/spree/customer_return_spec.rb:
```diff
@@ -0,0 +1,128 @@
+
+        it 'add an error message' do
+          customer_return.save!
+          expect(customer_return.errors.messages).to have_key(:shipwire)
```
do error messages have important information? can we also test their content?

kennyadsl commented on spec/models/spree/customer_return_spec.rb:
```diff
@@ -0,0 +1,128 @@
+      end
+
+
+      context 'when a return for the order was already been reported.)',
```
there's an extra `)` at the end of this line

kennyadsl commented on spec/models/spree/customer_return_spec.rb:
```diff
@@ -0,0 +1,128 @@
+        let(:return_item_1) { build(:return_item, inventory_unit: inventory_unit_1) }
+        let(:return_item_2) { build(:return_item, inventory_unit: inventory_unit_2) }
+        let(:shipwire_id)   { shipped_order_shipwire_id_2 }
+        let(:return_items)  { [return_item_1, return_item_1, return_item_2, return_item_2] }
```
can't you move these shared `let` outside the context so we avoid duplication? Or there's a reason to keep them duplicated?

kennyadsl commented on spec/models/spree/customer_return_spec.rb:
```diff
@@ -0,0 +1,128 @@
+        it 'will post to shipwire' do
+          expect(customer_return).to receive(:process_shipwire_return!).and_call_original
+          customer_return.save!
+          expect(customer_return.shipwire_id).not_to be nil
```
why in the previous content you used 2 scenarios and here just one? what's the difference?

kennyadsl commented on app/controllers/spree/admin/customer_returns_controller_decorator.rb:
```diff
@@ -0,0 +1,17 @@
+
+    private
+
+    def shipwire
```
can't we find a better name for this method?

aleph1ow committed:
Refactor class

aleph1ow committed:
Add a script to generate orders needed to specs

aleph1ow committed:
Fix customer return vcrs

aleph1ow committed:
Change method name

aleph1ow committed:
Improve return error handling

aleph1ow committed:
Refactor return errors handling

aleph1ow committed:
Fix returns error message default key

aleph1ow committed:
Post return after validation

we want to post a return after customer_return object validation
to eventually add errors to customer_return and inhibite save.

To make this possible we had to change the proxy create_on_shipwire method
to support objects (like CustomerReturn) that are not persisted before
the shipwire request is made.
So we added update_shipwire_id method that checks if the object is persisted
or not.

mtylty committed:
Merge remote-tracking branch 'upstream/master'

aleph1ow committed:
Fix migrations

aleph1ow committed:
Update cassette

aleph1ow committed:
Fix ruby syntax
